### ISSUE TITLE
New Block: Dark Mode Toggle

### ISSUE THREAD

**Issue by bhubbard** [WordPress/gutenberg#42828]
## What problem does this address?
A simple block to let site visitors toggle between dark mode and light mode.

**Comment by ramonjd** [comment]
This could be useful, but might also be tricky to deliver in the plugin because it would be very difficult to guarantee that it would work for all themes. 

How would you see it working, especially in terms of styles?

**Comment by bhubbard** [comment]
@ramonjd I could see the default being the toggle, if there are multiple styles available then I could see a possible setting to allow it to be a dropdown that lets the user choose the style.

**Comment by scruffian** [comment]
I think we'd need to add some tools for themes to be able to define "light" and "dark" colors.

**Comment by bhubbard** [comment]
> I think we'd need to add some tools for themes to be able to define "light" and "dark" colors.

@scruffian I agree, maybe this needs to be supported in theme.json for styles?

**Comment by bhubbard** [comment]
Any suggestions on how to move this forward?

**Comment by ramonjd** [comment]
Isn't this already possible with global style variations?

See: https://fullsiteediting.com/lessons/global-style-variations/

Given that light/dark modes will be very theme-specific, it makes sense that a theme could have 2 x global style variations to cater.

**Comment by bhubbard** [comment]
> Isn't this already possible with global style variations?
> 
> See: https://fullsiteediting.com/lessons/global-style-variations/
> 
> Given that light/dark modes will be very theme-specific, it makes sense that a theme could have 2 x global style variations to cater.

A theme can have many styles. The original request is to have a dark mode/light mode toggle button block, such as this one for example: https://www.w3schools.com/howto/howto_js_toggle_dark_mode.asp

This light mode/dark mode toggle is becoming more common across websites today. As a website visitor if my browser/os is set to dark mode, they may still wish to see the website in light mode. At the moment I don't think theme.json has a way to set a "style" as dark mode or light mode specific, nor does it support "prefer-color-scheme".

## Further Reading:
[Prefers Color Scheme](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)

**Comment by Lewiscowles1986** [comment]
Just a note that mainstream browsers have JS and CSS support for detecting the user preference on dark-mode. 

It's nothing but sparkling laziness to use a physical toggle if the default is not to respect the standards-based API's.

I'm also not 100% sold that it matters if the editing experience needs to be in lock-step with the display presentation theme. There is a preview button if someone needs to see what content looks like on their frontend as a visitor would see it.

I Created https://github.com/Automattic/blocks-everywhere/issues/27 for Gutenberg outside of WordPress. Which presently describes a hack, which "works for me". This hack is predicated on the visual editor not being a WYSIWYG website editor; which honestly neither TinyMCE or Gutenberg are capable of achieving despite a lot of sunk effort and painful engineering to that end.

**Comment by Lewiscowles1986** [comment]
Not sure who this bot is helping, but the progress on the linked issue I believe is that it was closed (marked as completed despite not being completed)
https://github.com/Automattic/blocks-everywhere/issues/27#issuecomment-1287577835
It's really simple; mark as want help, not priority and leave open as its honest about the shortcomings of the platform / tooling. All these bots producing sparkling nonsense does not help end-users.

**Comment by Lewiscowles1986** [comment]
This issue for example does not need any more information. The only labels applied that make sense are `[Type] Enhancement`.

I'd argue that from [labels](https://github.com/WordPress/gutenberg/labels), the following seemed like they might apply

- [Accessibility (a11y)](https://github.com/WordPress/gutenberg/labels/Accessibility%20%28a11y%29)
- [[a11y] Color Contrast](https://github.com/WordPress/gutenberg/labels/%5Ba11y%5D%20Color%20Contrast)
- [CSS Styling](https://github.com/WordPress/gutenberg/labels/CSS%20Styling)
- [[Feature] Colors](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Colors)
- [[Feature] Custom Editor Styles](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Custom%20Editor%20Styles)
- [[Feature] Design Tools](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Design%20Tools)
- [General Interface](https://github.com/WordPress/gutenberg/labels/General%20Interface)
- [[Feature] Style Variations](https://github.com/WordPress/gutenberg/labels/%5BFeature%5D%20Style%20Variations)
- [[Priority] Low](https://github.com/WordPress/gutenberg/labels/%5BPriority%5D%20Low)
- [[Status] Blocked](https://github.com/WordPress/gutenberg/labels/%5BStatus%5D%20Blocked) or [[Status] Not Implemented](https://github.com/WordPress/gutenberg/labels/%5BStatus%5D%20Not%20Implemented)

**Comment by bhubbard** [comment]
Maybe we can move this forward by adding support for "prefers-color-scheme" to the styles JSON? 

prefers-color-scheme: dark
prefers-color-scheme: light

- https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme
- https://github.com/WordPress/gutenberg/issues/24368
- https://github.com/WordPress/gutenberg/issues/16826

**Comment by Lewiscowles1986** [comment]
Thanks @bhubbard 

So one limitation of a solely CSS based approach (although that is better than what we have now); is that it does not let a user set an override (for session or otherwise).

I did use the prefers-color-scheme, by-proxy, but because the software I'd integrated Gutenberg with had system theme independence, I also used some JS https://github.com/Automattic/blocks-everywhere/issues/27

I Think not locking down to solely CSS is an ideal world, and I'm open to any world where I'm not writing custom CSS hacks 👍

**Comment by Lewiscowles1986** [comment]
I'd like to move that bot forward... into the bin. While I understand the need to keep issues fresh and assess interest, it comes across as disinterest. 

- https://blogs.perficient.com/2022/03/04/dark-mode-and-digital-accessibility-who-it-helps-and-without-it-who-it-may-hinder/
- https://www.nngroup.com/articles/dark-mode/ (benefits for users with cataracts)

For me it was purely aesthetic, but there are always benefits to supporting multi-palette UI if the cost can be controlled.

**Comment by bhubbard** [comment]
@richtabor made a block for this: https://rich.blog/dark-mode-toggle-block/

**Comment by Lewiscowles1986** [comment]
So is this asking for a block for websites to add a toggle, or for Gutenberg itself to support dark-mode?

All my posts were for the latter, not the former.

**Comment by bhubbard** [comment]
@Lewiscowles1986 So the original request is for a new toggle block. Browsing a lot of templates, I do see a lot of sites doing a drop-down that has light, dark, and system (default) as the options. And I feel this is probably best, as it defaults to system, but site visitors can then choose to see light and dark when they wish.

Example:
![fb5a636f-f605-4ca3-b224-8a07c63d9912](https://github.com/WordPress/gutenberg/assets/3837/b69247cf-11b8-45f3-a5cc-82b2eed3873c)

Based on the comments, I do see some work that may need to be completed first, such as let developers define "prefers-color-scheme" for color palette's set within the theme.json file. It might also make sense to have this as a site option stored in the wp_options table, with default being system, but allowing site owners to override the default for the site globally.

**Comment by Lewiscowles1986** [comment]
@bhubbard are wp_options user-specific?
I thought it was not https://codex.wordpress.org/Database_Description#Table:_wp_options

To add complexity, I'm not 100% certain, that a user wanting dark-mode, would wish to do so for all of their devices.

**Comment by bhubbard** [comment]
@Lewiscowles1986 wp_options is not user-specific. But I was thinking its more for site owners to choose the default for their site, if they choose to not have it be "system". The idea being a user can always use this toggle or drop-down to then change on the front-end of the site.

**Comment by annezazu** [comment]
Thanks for this discussion here. In reviewing this in light of a push for new blocks in Core https://github.com/WordPress/gutenberg/issues/71026, I'm going to close this out as a duplicate of a broader issue here with more discussion https://github.com/WordPress/gutenberg/issues/48249 As you can see in that discussion too, there is not a clear way forward nor easy solution for this to core. For now, check out https://developer.wordpress.org/news/2025/09/building-a-light-dark-toggle-with-the-interactivity-api/ which offers a way to do this for theme authors.

**Comment by Lewiscowles1986** [comment]
@annezazu how can 42828 be a duplicate of 48249; the latter, was created nearly 1 year later.

Thank you for linking it. As I see it, at least what I was hoping for wasn't captured in this issue or the new one.

**Comment by annezazu** [comment]
Great question! When triaging and noticing similar issues, a judgement has to be made around what to keep open. Sometimes this means keeping open an issue that was created later but that, for whatever reason, has more traction and discussion. When that happens, in order to better preserve where momentum and context is (including the number of folks who’ve engaged and are following the issue), we’ll close the older issue. It’s ultimately a judgment call. 

With that context in mind, what is missing from the other issue? I see a clear reference to having a high contrast and dark mode.

**Comment by Lewiscowles1986** [comment]
Thank you, for your answer, and patience.

The other ticket is about implementing a block for end-users controlling the dark / light theme, as I believe this one may have originally been about.

Missing for that use case would be respecting the system theme and what current or future wordpress mechanisms to leverage for that.

I Should also create a separate issue for dark mode within the backend Gutenberg experience, but I don't know if there is any desire beyond my own for it; and I have a hack that gets me dark-mode looking Gutenberg.

**Comment by annezazu** [comment]
Of course. Thanks for engaging so kindly. 

>I Should also create a separate issue for dark mode within the backend Gutenberg experience, but I don't know if there is any desire beyond my own for it; and I have a hack that gets me dark-mode looking Gutenberg.

Oh! At a broad level, this is being thought of right now with upcoming Admin Design efforts. Check out the parts around "Theming": https://github.com/WordPress/gutenberg/issues/71196

>Missing for that use case would be respecting the system theme and what current or future wordpress mechanisms to leverage for that.

I think this is where there's not a clear path forward for this as sites aren't applications and would likely need to be handled on a per site basis. It's complex, that's for sure.
